protocol/transport.go

```diff
@@ -86,21 +100,31 @@ func (t *Transport) Read() (msg Message, err error) {
 		if err != nil {
 			return
 		}
-		msg, err = t.read()
+		msg, err = t.readFrontendMessage()
 	}
 	if err != nil {
 		return
 	}
 
-	if msg.CreatesTransaction() && t.transaction == nil {
-		t.beginTransaction()
-	} else if msg.EndsTransaction() && t.transaction != nil {
-		err = t.endTransaction()
+	if t.transaction == nil {
+		switch msg.(type) {
+		case *pgproto3.Parse, *pgproto3.Bind, *pgproto3.Describe:
+			t.beginTransaction()
+		}
+	} else {
+		switch msg.(type) {
+		case *pgproto3.Query, *pgproto3.Sync:
+			err = t.endTransaction()
+		}
 	}
 
 	return
 }
 
+func (t *Transport) readFrontendMessage() (pgproto3.FrontendMessage, error) {
+	return t.backend.Receive()
```

Comment:
🤔 
super confusing that we use t.backend to get next front message..
what do you think about calling it `frontReader *pgproto3.Backend` or something similar?